### PR TITLE
Feaure lwdev 9227 cashout aggregation

### PR DESCRIPTION
**Pull request by OttoVT**

### CODE REVIEW

File: src/Lykke.Service.History/Workflow/Projections/TransactionHashProjection.cs [hunk 1]
```diff
@@ -87,5 +87,31 @@ public async Task<CommandHandlingResult> Handle(CashoutCompletedEvent @event)
...
+        ///     BIL batched cashout event
+        /// </summary>
+        /// <param name="event"></param>
+        /// <returns></returns>
+        public async Task<CommandHandlingResult> Handle(
+            Job.BlockchainCashoutProcessor.Contract.Events.CashoutsBatchCompletedEvent @event)
+        {
+            if (@event.Cashouts == null || @event.Cashouts.Length == 0)
+            {
+                _logger.Warning($"BIL batched cashout event, BatchId {@event.BatchId}", context: @event);
```
Comment (some-dev) [inline src/Lykke.Service.History/Workflow/Projections/TransactionHashProjection.cs]: so if Cashouts == null - NRE is throwed below in "foreach" clause. We should add "return" state after line with logging.

Comment (OttoVT) [reply src/Lykke.Service.History/Workflow/Projections/TransactionHashProjection.cs]: thx!

**Review by some-dev — DISMISSED** [review]

File: src/Lykke.Service.History/Workflow/Projections/TransactionHashProjection.cs [hunk 2]
```diff
@@ -87,5 +87,33 @@ public async Task<CommandHandlingResult> Handle(CashoutCompletedEvent @event)
...
+
+                return CommandHandlingResult.Ok();
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.History/Workflow/Projections/TransactionHashProjection.cs:121]: This is not OK, throw an exception in this case

File: src/Lykke.Service.History/Modules/CqrsModule.cs [hunk 3]
```diff
@@ -144,7 +144,9 @@ protected override void Load(ContainerBuilder builder)
                     .WithEndpointResolver(sagasMessagePackEndpointResolver)
                     .WithProjection(typeof(TransactionHashProjection), BlockchainCashinDetectorBoundedContext.Name)
 
-                    .ListeningEvents(typeof(Job.BlockchainCashoutProcessor.Contract.Events.CashoutCompletedEvent))
+                    .ListeningEvents(
+                        typeof(Job.BlockchainCashoutProcessor.Contract.Events.CashoutCompletedEvent),
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.History/Modules/CqrsModule.cs:148]: Where is CrossClientCashoutCompletedEvent missed?

File: src/Lykke.Service.History/Workflow/Projections/TransactionHashProjection.cs [hunk 4]
```diff
@@ -87,5 +87,33 @@ public async Task<CommandHandlingResult> Handle(CashoutCompletedEvent @event)
...
+            }
+
+            foreach (var cashout in @event.Cashouts)
+            {
+                if (!await _historyRecordsRepository.UpdateBlockchainHashAsync(cashout.OperationId, @event.TransactionHash))
+                    _logger.Warning($"Transaction hash was not set, BIL cashout", context: new
```
Comment (KonstantinRyazantsev) [inline src/Lykke.Service.History/Workflow/Projections/TransactionHashProjection.cs:129]: This means that some data is in not consistent state, we should throw an exception to retry, not just log a warning

**Review by KonstantinRyazantsev — DISMISSED** [review]

File: src/Lykke.Service.History/Workflow/Projections/TransactionHashProjection.cs [hunk 5]
```diff
@@ -26,11 +28,8 @@ public TransactionHashProjection(IHistoryRecordsRepository historyRecordsReposit
         public async Task<CommandHandlingResult> Handle(CashinCompletedEvent @event)
         {
             if (!await _historyRecordsRepository.UpdateBlockchainHashAsync(@event.OperationId, @event.TxHash))
-                _logger.Warning($"Transaction hash was not set, bitcoin cashin", context: new
-                {
-                    id = @event.OperationId,
-                    hash = @event.TxHash
-                });
+                throw new InvalidOperationException($"Transaction hash was not set. " +
```
Comment (polonzer) [inline src/Lykke.Service.History/Workflow/Projections/TransactionHashProjection.cs]: Why throwing exception? It seems like log.Warning + Retry should be the right choice here. The same applies to the code below.

**Review by polonzer — DISMISSED** [review]

**Review by KonstantinRyazantsev — APPROVED** [review]

**Pull request merged**